README
This application will allow users of Django-powered websites keep track of the time they spend on different projects.  It's a work in process.

==Features==

  * Configurable: Pendulum can be configured to operate on several Django-powered sites.  The period lengths can be configured as monthly or as a fixed-length period.
  * Projects: You can have an unlimited number of projects to be able to categorize hours spent working on certain tasks.  Each project can be activated/deactivated as necessary via the Django admin.
  * Activities: Activities allow you to further categorize work done on particular tasks for each project.

==Requirements==

`django-timepiece` requires a modern version of the Django framework.  By modern I simply mean a version with the `newforms-admin` functionality.  If you're running on Django 1.0 or later, you're good.

`django-timepiece` also relies upon the `django.contrib.admin`, `django.contrib.auth`, `django.contrib.contenttypes`, and `django.contrib.sites` frameworks.

==Installation==

Download `django-timepiece` using *one* of the following methods:

===easy_install===

You can download the package from the [http://pypi.python.org/pypi/django-timepiece/ CheeseShop] or use

{{{
easy_install django-timepiece
}}}

to download and install `django-timepiece`.

===Package Download===

Download the latest `.tar.gz` file from the downloads section and extract it somewhere you'll remember.  Use `python setup.py install` to install it.

===Checkout from Subversion===

Execute the following command (or use the equivalent function in a GUI such as TortoiseSVN), and make sure you're checking Pendulum out somewhere on the `PYTHONPATH`.

{{{
svn co http://django-timepiece.googlecode.com/svn/trunk/timepiece timepiece
}}}

===Verifying Installation===

The easiest way to ensure that you have successfully installed Pendulum is to execute a command such as:

{{{
python -c "import timepiece; print timepiece.get_version()"
}}}

If that displays the version of Pendulum that you tried to install, you're good to roll.  If you see something other than that, you probably need to check your `PYTHONPATH` environment variable.

==Configuration==

First of all, you must add this project to your list of `INSTALLED_APPS` in `settings.py`:

{{{
INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    ...
    'timepiece',
    ...
)
}}}

Run `manage.py syncdb`.  This creates a few tables in your database that are necessary for operation.

Next, you should add an entry to your main `urls.py` file.  For example:

{{{
from django.conf.urls.defaults import *

from django.contrib import admin
admin.autodiscover()

urlpatterns = patterns('',
    (r'^admin/(.*)', admin.site.root),
    (r'^timepiece/', include('timepiece.urls')),
)
}}}

The next thing you will want to do is configure Pendulum for the active Django sites.  Do this by going into the Django admin and clicking the "add" link next to Pendulum Configurations.

The first step in the configuration is to choose which site this particular configuration will apply to.  The decision is easy if you only have one site :)

Next, you must choose what kind of "accounting period" you wish to use.  If you want month-long periods, ensure the "Is monthly" check box is selected.  Then enter the day of the month that the periods begin.  Periods that begin on the first of the month will always range from midnight on the first of each month to 11:59:59 PM on the last day of each month.  Periods that begin on the 16th of the month will range from midnight on the 16th of each month to 11:59:59 PM on the 15th of the following month.  Be wise about choosing start days such as the 31st of the month--it's not a wise choice no matter what situation you're in.

Alternatively, you may choose fixed-length accounting periods.  The installation date does not necessarily have to be the actual installation date.  This field is used as a reference point for creating periods.  For example, if you have 2-week long accounting periods that begin on Sundays and end two Saturdays later, you would select _any_ Sunday from any month from any year.  Then you would enter 14 for the period length.  Periods will automatically be constructed that will begin at midnight each Sunday and end at 11:59:59 PM two Saturdays in the future.

If you fail to configure Pendulum for a given site, you will likely receive HTTP 500 errors if you try to access Pendulum via the front-end.

===Add Projects===

Next, you should add at least one project to Pendulum.  This can easily be accomplished via the Django admin.  The same goes for activities, but these are not required.

===Permissions===

Finally, be sure set people up with the appropriate permissions using Django's Auth framework.  The front end side of Pendulum respects these permissions.  That means that if a user doesn't have the permission to clock in, the will be presented with a login screen (even if they're logged in already) when they try to clock in on the front end site.  The same goes for other operations such as pausing/unpausing entries, updating entries, deleting entries, etc.

An easy way to assign the same Pendulum permissions to several users is to create a Group using the Auth framework.  There is a custom command for creating a group with all of the permissions necessary for the front end, including the ability to:

  * add entries
  * update entries
  * delete entries
  * clock in
  * clock out
  * pause/unpause entries

To use this command, simply run `./manage.py create_timepiece_group`.  You may run `./manage.py help create_timepiece_group` to learn about options for this command.

==The Date Field==

I've been hesitant to do anything to make the date/time fields more user-friendly when adding or updating entries.  I don't want to tie anyone down to my particular way of doing things.  I personally prefer jQuery for this sort of thing, but I do realize that jQuery disgusts many other people out there.  Therefore, the first versions of the application didn't come with anything to make those fields easier to use and understand.

However, that seems to be the most common complaint amongst those who have played with the application--the lack of user-friendly date pickers.  As such, I have added some jQuery-based utilities to make choosing a date easier and more intuitive.

By default, the application will expect to find the appropriate files in a directory called `timepiece` within your media directory.  All you should need to do is copy or symlink the `media` folder from the Pendulum application directory.  If you copy the `media` folder, rename it to `timepiece`.  If you create a symlink to Pendulum's `media` folder, make sure the symlink is called `timepiece`.  This directory should hold all of the necessary files to make the date picker work.

In your `add_update_entry.html` template (if you override the default ones), make sure you have a code block such as:

{{{
{% block extra-head %}
{{ form.media }}
<script type="text/javascript">
$(document).ready(function () {
    $('input.vDateField').datepicker({
            dateFormat: $.datepicker.W3C,
            showOn: 'both',
            buttonImage: '{{ MEDIA_URL }}timepiece/img/calendar.png',
            buttonImageOnly: true
        });
});
</script>
{% endblock %}
}}}

This is just jQuery's way of attaching the date picker to the appropriate fields.

===Customizing The Date Picker===

I tried to put some utilities in the application to allow you to change the date picker that is used with ease.  There is a variable that you can define in your `settings.py` file to override the default date picker: `PENDULUM_DATE_MEDIA`.  If you simply want to change the styling CSS,, for example, you could use this variable.  The default value is:

{{{
PENDULUM_DATE_MEDIA = {
    'js': (settings.MEDIA_URL + 'timepiece/js/jquery.js',
            settings.MEDIA_URL + 'timepiece/js/jquery.ui.js'),
    'css': {
        'all': (settings.MEDIA_URL + 'timepiece/css/jquery-ui.css',)
    }
}
}}}

You can see that it is a dictionary with two keys: `js` and `css`.  The values of these two items conform to the concepts outlined in [http://docs.djangoproject.com/en/dev/topics/forms/media/ Django's documentation].  Changing the values within `PENDULUM_DATE_MEDIA` should permit you to completely remove or replace the date picker as you desire.


docs/releasenotes.rst
Release Notes
=============

1.1.0 (2016-02-29)
----------------------------

*Bugfixes*

* Activity autocomplete on entry creation / editing now works. Activity search
results were previously not actually being restricted by the search term, and
incorrect handling of empty project values was causing errors when the user
started to search for an activity without selecting a project.
* Invoice totals now reflect the summing up of values rounded to two decimal
places rather than the rounding to two decimal places of a sum of numbers
with five decimal places.

1.0.0 (2016-01-04)
----------------------------

Release 1.0.0 is a comprehensive update of the django-timepiece codebase. It is
based on 0.9.5, which was not officially released to PyPI. Release
1.0.0 includes the bulk of the unofficial 0.9.5 release, along with several critical
code quality improvements to 0.9.5.

Please note that **release 1.0.0 is not backwards compatible with any 0.9.x release.**
It is a new and separate refactoring of django-timepiece.

*Features*

* Assigning Project hours to a User not associated with the Project will automatically
associate User with Project
* Admins are now able to add Entries after User's timesheet has been approved
* Payroll summary now groups by User last name
* Dashboard comments have been converted to modal popups to keep interface clean
* Contract detail now shows total hours worked before Contract start date and post
Contract end date
* Sticky headers have been added to all tables
* Time Entries are now more precise to prevent the number of seconds being off in totals
* Tabs for pending and completed Projects in the Contracts page
* Users can only log activities that are allowed for a Project
* Estimation Accuracy report has been refined for readability

*Bugfixes*

* Alphabetized entries in weekly schedule
* Previous/Next options returned to the bottom of pages
* Hourly Report outputs correct data
* Sum of parts now accurately matches title (rounding display has been removed to show explicit
times)
* Dashboard header styling has been made consistent (hover styling removed and background color
matched)
* lessc installs locally rather than globally

*Code Quality*

* Compatibility with Django 3.2 with Python 3.8
* Drops support for all other combinations of Django and Python
* Drops support for testing with Jenkins
* Flake8 compliance
* Tests updated to clean up stray data which led to inconsistent failures

0.9.5 (Not Released to PyPI)
----------------------------

*Features*

* Use Bootstrap 2.3.2
* Estimation Accuracy report
* Show active entry comment on dashboard page.
* Order previous invoices by end date descending.

*Bugfixes*

* Save auth groups when creating a new user.
* Require `crm.change_business` permission rather than `crm.edit_business`
  permission for EditUser view.
* Don't use transaction.commit_on_success on ViewUser or ViewProject views.
* Fix bug preventing passage of custom default value to `utils.get_setting`.
* Manually calculate height/width in hours (#147b95a).
* Fixed migrations so that they will run on a fresh database. If you have
  an existing database, you will need to run ``django-admin.py migrate
  entries 0002 --fake``.

*Code Quality*

* Additional tests and code cleanup in the CRM app.
* Use Model.add_to_class to augment class.
* Make use of @cbv_decorator to avoid creating new Mixin classes for every
  decorator.
* Reorganized navbar templates to make extension easier when Timepiece is a
  secondary app. Thanks @dannybrowne86!

0.9.4 (Released 04-01-2014)
---------------------------

* Click to highlight row/column on Payroll summary page

0.9.3 (Released 11-06-2013)
---------------------------

Related issues are in the `0.9.3. milestone
<https://github.com/caktus/django-timepiece/issues?milestone=39&state=closed>`_.

*Features*

* Allow using compress when `DEBUG = True` with a new context processor,
  `timepiece.context_processors.extra_settings`.
* Align daily summary tables on the user timesheet
  (`#725 <https://github.com/caktus/django-timepiece/pull/725>`_).
* Add progress bars on the contract listing page
  (`#707 <https://github.com/caktus/django-timepiece/pull/707>`_).
* Add vertical highlighting to the payroll summary report
  (`#727 <https://github.com/caktus/django-timepiece/issues/727>`_).
* Bump `django-selectable <https://django-selectable.readthedocs.org/en/version-0.7.0/>`_
  version number to 0.7.0.
* Added a warning on the outstanding invoices page if users have unverified/unapproved
  entries for the selected time period (`#744 <https://github.com/caktus/django-timepiece/pull/744>`_).

*Bugfixes*

* Include entries on last date of query on invoice listing page
  (`#718 <https://github.com/caktus/django-timepiece/issues/718>`_).
* Display correct user name on the timesheet rejection confirmation page
  (`#726 <https://github.com/caktus/django-timepiece/issues/726>`_).
* Register `contracts.ContractHour` in the admin so that the
  `get_absolute_url()` method can work.

*Code Quality*

* Started refactoring and improving the test suite. `factory_boy
  <https://factoryboy.readthedocs.org/en/latest/>`_ is now required to run the
  tests.
* Refactored most CRM views to be class-based, and added more tests.
* Added several CBV mixins to `utils.mixins`.


0.9.2 (Released 05-31-2013)
---------------------------

Related issues are in the `0.9.2 milestone
<https://github.com/caktus/django-timepiece/issues?milestone=34&state=closed>`_.

*Features*

* Checkbox filter by status on the outstanding hours list view.
* Click-to-highlight rows on the payroll summary page.
* Administrators with 'entries.add_projecthours' permission can view
  unpublished hours on the schedule view page.
* Refactored Total Progress area on the dashboard to more clearly show
  total hours worked vs. total hours assigned.
* Aligned columns on time sheet and dashboard All Entries tables.
* Displayed negative times with a negative sign and a red font.

pendulum/models.py
import datetime

from django.db import models
from django.contrib.sites.models import Site
from django.contrib.auth.models import User

from pendulum import utils

from dateutil.relativedelta import relativedelta


try:
    """
    if the app is installed from the start of a site, the sites table does
    not exist and it causes problems.  This exception block seems to fix
    the problem.
    """
    CURRENT_SITE = Site.objects.get_current()
except:
    CURRENT_SITE = Site.objects.all()

class PendulumConfiguration(models.Model):
    """
    This will hold a single record that maintains the configuration of the
    application.  In the admin interface, if the configuration is marked as
    "Is Monthly", that will take precedence over the "Install date" option (even
    if the install_date and period_length fields have values).  If you wish to
    use the fixed-length (install_date + period_length) setup, uncheck the
    is_monthly field.
    """

    # tie the configuration to one site in particular
    site = models.OneToOneField(Site, help_text="""Please choose the site that these settings will apply to.""")

    """
    this represents whether the application will look for all entries in a
    month-long period
    """
    is_monthly = models.BooleanField(default=True, help_text="""If you check this box, you will be forced to use the monthly mode.  Uncheck it to use fixed-length period""")

    """
    this is used in conjunction with the monthly setup; end date is assumed
    to be month_start - 1.  For example, if the periods begin on the 16th of
    each month, the end date would be assumed to be the 15th of each month
    """
    month_start = models.PositiveIntegerField(default=1, blank=True, null=True,
                                              help_text="""Enter 1 for periods that begin on the 1st day of each month and end on the last day of each month.  Alternatively, enter any other number (between 2 and 31) for the day of the month that periods start.  For example, enter 16 for periods that begin on the 16th of each month and end on the 15th of the following month.""")

    """
    install_date represents the date the software was installed and began
    being used.  period_length represents the number of days in a period.  Week-
    long periods would have a period_length of 7.  Two week-long periods would
    be 14 days.  You get the idea.  These should be able to handle _most_
    situations (maybe not all).
    """
    install_date = models.DateField(blank=True, null=True, help_text="""The date that Pendulum was installed.  Does not necessarily have to be the date, just a date to be used as a reference point for adding the number of days from period length below.  For example, if you have periods with a fixed length of 2 weeks, enter 14 days for period length and choose any Sunday to be the install date.""")
    period_length = models.PositiveIntegerField(blank=True, null=True, help_text="""The number of days in the fixed-length period.  For example, enter 7 days for 1-week periods or 28 for 4-week long periods.""")

    def __unicode__(self):
        return u'Pendulum Configuration for %s' % self.site

    def __current_mode(self):
        if self.is_monthly:
            return u'Month-long'
        else:
            return u'Fixed-length'
    current_mode = property(__current_mode)

class ProjectManager(models.Manager):
    """
    Return all active projects.
    """
    def get_query_set(self):
        return super(ProjectManager, self).get_query_set().filter(sites__exact=CURRENT_SITE)

    def active(self):
        return self.get_query_set().filter(is_active=True)

class Project(models.Model):
    """
    This class will keep track of different projects that one may clock into
    """

    name = models.CharField(max_length=100, unique=True,
                            help_text="""Please enter a name for this project.""")
    description = models.TextField(blank=True, null=True,
                                   help_text="""If necessary, enter something to describe the project.""")
    is_active = models.BooleanField(default=True,
                                    help_text="""Should this project be available for users to clock into?""")
    sites = models.ManyToManyField(Site, related_name='pendulum_projects',
                                  help_text="""Choose the site(s) that will display this project.""")
    date_added = models.DateTimeField(auto_now_add=True)
    date_updated = models.DateTimeField(auto_now=True)

    objects = ProjectManager()

    def __unicode__(self):
        """
        The string representation of an instance of this class
        """
        return self.name

    def log_count(self):
        """
        Determine the number of entries associated with this project
        """
        return self.entries.all().count()

    def __total_hours(self):
        """
        Determine the number of hours spent working on each project
        """
        times = [e.total_hours for e in self.entries.all()]
        return '%.02f' % sum(times)
    total_hours = property(__total_hours)

    class Meta:
        ordering = ['name', 'date_added']

class ActivityManager(models.Manager):
    """
    Return all active activities.
    """
    def get_query_set(self):
        return super(ActivityManager, self).get_query_set().filter(sites__exact=CURRENT_SITE)

class Activity(models.Model):
    """
    Represents different types of activity: debugging, developing,
    brainstorming, QA, etc...
    """

    code = models.CharField(max_length=5, unique=True,
                            help_text="""Enter a short code to describe the type of activity that took place.""")
    name = models.CharField(max_length=50,
                            help_text="""Now enter a more meaningful name for the activity.""")
    sites = models.ManyToManyField(Site, related_name='pendulum_activities',
                                  help_text="""Choose the site(s) that will display this activity.""")

    objects = ActivityManager()

    def __unicode__(self):
        """
        The string representation of an instance of this class
        """
        return self.name

    def __log_count(self):
        """
        Determine the number of entries associated with this activity
        """
        return self.entries.all().count()
    log_count = property(__log_count)

    def __total_hours(self):
        """
        Determine the number of hours spent doing each type of activity
        """
        times = [e.total_hours for e in self.entries.all()]
        return '%.02f' % sum(times)
    total_hours = property(__total_hours)

    class Meta:
        ordering = ['name']
        verbose_name_plural = 'activities'

class EntryManager(models.Manager):
    #def get_query_set(self):
    #    return super(EntryManager, self).get_query_set().filter(site__exact=CURRENT_SITE)

    def current(self, user=None):
        """
        This will pull back any log entries for the current period.
        """
        try:
            set = self.in_period(utils.determine_period())
        except PendulumConfiguration.DoesNotExist:
            raise Exception, "Please configure Pendulum!"
        else:
            if user:
                return set.filter(user=user)
            return set

    def previous(self, delta, user=None):
        set = self.in_period(utils.determine_period(delta=delta))

        if user:
            return set.filter(user=user)
        return set

    def in_period(self, period, user=None):
        if not isinstance(period, tuple) or len(period) != 2:
            raise Exception('Invalid period specified')

        set = self.get_query_set().filter(start_time__range=period)

        if user:
            return set.filter(user=user)
        return set

class Entry(models.Model):
    """
    This class is where all of the time logs are taken care of
    """

    user = models.ForeignKey(User, related_name='pendulum_entries')
    project = models.ForeignKey(Project,
                                limit_choices_to={'is_active': True,
                                                  'sites': CURRENT_SITE},
                                related_name='entries')
    activity = models.ForeignKey(Activity, blank=True, null=True, related_name='entries')
    start_time = models.DateTimeField()
    end_time = models.DateTimeField(blank=True, null=True)
    seconds_paused = models.PositiveIntegerField(default=0)
    pause_time = models.DateTimeField(blank=True, null=True)
    comments = models.TextField(blank=True, null=True)
    date_updated = models.DateTimeField(auto_now=True)
    site = models.ForeignKey(Site, related_name='pendulum_entries')
    hours = models.DecimalField(max_digits=8, decimal_places=2)

    objects = EntryManager()

    def get_seconds(self):
        """
        Determines the difference between the starting and ending time.  The
        result is returned as an integer of seconds.
        """
        if self.start_time and self.end_time:
            # only calculate when the start and end are defined
            delta = self.end_time - self.start_time
            seconds = delta.seconds - self.seconds_paused
        else:
            seconds = 0
            delta = timedelta(days=0)

        return seconds + (delta.days * 86400)

    def __total_hours(self):
        """
        Determined the total number of hours worked in this entry
        """
        return self.get_seconds() / 3600.0
    total_hours = property(__total_hours)

    def __total_time(self):
        """
        Determines the amount of time spent and return it as a string formatted
        as HH:MM:SS
        """
        return utils.get_total_time(self.get_seconds())
    total_time = property(__total_time)

    def __paused_time(self):
        """
        Returns the total time paused for this entry in HH:MM:SS format
        """
        return utils.get_total_time(self.seconds_paused)
    paused_time = property(__paused_time)

    def __hours(self):
        """
        Print the hours in a nice, rounded format
        """
        return "%.02f" % self.total_hours
    nice_hours = property(__hours)

    def __is_paused(self):
        """
        Determine whether or not this entry is paused
        """
        return self.pause_time != None
    is_paused = property(__is_paused)

    def pause(self):
        """
        If this entry is not paused, pause it.
        """
        if not self.is_paused:
            self.pause_time = datetime.datetime.now()

    def unpause(self):
        """
        If this entry is paused, unpause it
        """
        if self.is_paused:
            delta = datetime.datetime.now() - self.pause_time
            self.seconds_paused += delta.seconds
            self.pause_time = None

    def toggle_paused(self):
        """
        Toggle the paused state of this entry.  If the entry is already paused,
        it will be unpaused; if it is not paused, it will be paused.
        """
        if self.is_paused:
            self.unpause()
        else:
            self.pause()

    def __is_closed(self):
        """
        Determine whether this entry has been closed or not
        """
        return self.end_time != None
    is_closed = property(__is_closed)

    def clock_in(self, user, project):
        """
        Set this entry up for saving the first time, as an open entry.
        """
        if not self.is_closed:
            self.user = user
            self.project = project
            self.site = CURRENT_SITE
            self.start_time = datetime.datetime.now()

    def clock_out(self, activity, comments):
        """
        Save some vital pieces of information about this entry upon closing
        """
        if self.is_paused:
            self.unpause()

        if not self.is_closed:
            self.end_time = datetime.datetime.now()
            self.activity = activity
            self.comments = comments

    def __delete_key(self):
        """
        Make it a little more interesting for deleting logs
        """
        salt = '%i-%i-apple-%s-sauce' % (self.id, self.is_paused, self.is_closed)
        try:
            import hashlib
        except ImportError:
            import sha
            key = sha.new(salt).hexdigest()
        else:
            key = hashlib.sha1(salt).hexdigest()
        return key
    delete_key = property(__delete_key)

    def __unicode__(self):
        """
        The string representation of an instance of this class
        """
        return '%s on %s' % (self.user, self.project)

    class Meta:
        # ordering = ['-start_time']
        verbose_name_plural = 'entries'
        permissions = (
            ('can_clock_in', 'Can use Pendulum to clock in'),
            ('can_pause', 'Can pause and unpause log entries'),
            ('can_clock_out', 'Can use Pendulum to clock out'),
        )

# Add a utility method to the User class that will tell whether or not a
# particular user has any unclosed entries
User.clocked_in = property(lambda user: user.pendulum_entries.filter(end_time__isnull=True).count() > 0)


class RepeatPeriod(models.Model):
    INTERVAL_CHOICES = (
        ('day', 'Day(s)'),
        ('week', 'Week(s)'),
        ('month', 'Month(s)'),
        ('year', 'Year(s)'),
    )
    project = models.ForeignKey(Project, unique=True)
    count = models.PositiveSmallIntegerField(
        # null=True,
        # blank=True,
        choices=[(x,x) for x in range(1,32)],
    )
    interval = models.CharField(
        # null=True,
        # blank=True,
        max_length=10,
        choices=INTERVAL_CHOICES,
    )
    active = models.BooleanField(default=False)
    
    def __unicode__(self):
        return "%d %s for %s" % (self.count, self.get_interval_display(), self.project)
    
    def delta(self):
        return relativedelta(**{str(self.interval + 's'): self.count})
    
    def update_billing_windows(self):
        windows = []
        try:
            window = self.billing_windows.order_by('-date').select_related()[0]
        except IndexError:
            window = None
        if window:
            while window.date + self.delta() <= datetime.date.today():
                window.id = None
                window.date += self.delta()
                window.end_date += self.delta()
                window.save(force_insert=True)
                windows.append(window)
        else:
            window = self.billing_windows.create(
                date=datetime.date.today(),
                end_date=datetime.date.today() + self.delta(),
            )
            windows.append(window)
        return windows
    
    def get_window(self, window_id):
        try:
            window = self.billing_windows.select_related().get(pk=window_id)
        except BillingWindow.DoesNotExist:
            window = None
        return window
    
    def get_latest_window(self):
        try:
             window = self.billing_windows.select_related().latest()
        except BillingWindow.DoesNotExist:
            window = None
        if not window or window.date + window.period.delta() <= datetime.date.today():
            window = self.update_billing_windows()[-1]
        return window


class BillingWindow(models.Model):
    period = models.ForeignKey(RepeatPeriod, related_name='billing_windows')
    date = models.DateField()
    end_date = models.DateField()
    
    class Meta:
        get_latest_by = 'date'
    
    def __unicode__(self):
        return "%s through %s" % (self.date, self.end_date)
    
    def get_entries(self):
        if not hasattr(self, '_entries'):
            self._entries = Entry.objects.filter(
                project=self.period.project,
                start_time__gte=self.date,
                end_time__lt=self.end_date,
            ).select_related().order_by('start_time')
        return self._entries
    
    def total_hours(self):
        return Entry.objects.filter(
            project=self.period.project,
            start_time__gte=self.date,
            end_time__lt=self.end_date,
        ).aggregate(hours=models.Sum('hours'))['hours']
    
    def next(self):
        if not hasattr(self, '_next'):
            try:
                window = BillingWindow.objects.filter(
                    period=self.period,
                    date__gt=self.date,
                ).order_by('date')[0]
            except IndexError:
                window = None
            self._next = window
        return self._next
    
    def previous(self):
        if not hasattr(self, '_previous'):
            try:
                window = BillingWindow.objects.filter(
                    period=self.period,
                    date__lt=self.date,
                ).order_by('-date')[0]
            except IndexError:
                window = None
            self._previous = window
        return self._previous
